### PR TITLE
HPCC-19393 Implement single-use security manager plugin

### PR DESCRIPTION
**Pull request by rpastrana**
- User/pass provided in config by the espcfg component
- User/pass must be provided to ESP.exe via command line params
- Defaults user to "admin" if omitted

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by rpastrana** [comment]
@wangkx please review.

File: system/security/plugins/singleuserSecurity/singleUserSecurity.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+#include "basesecurity.hpp"
+#include "authmap.ipp"
+#include "singleUserSecurity.hpp"
+
+class CSingleUserSecurityManager : public CBaseSecurityManager
+{
+public:
+    CSingleUserSecurityManager(const char *serviceName, IPropertyTree *secMgrCfg, IPropertyTree *bindConfig) : CBaseSecurityManager(serviceName, (IPropertyTree *)NULL)
+    {
+        if (secMgrCfg)
```
Comment (wangkx) [inline system/security/plugins/singleuserSecurity/singleUserSecurity.cpp]: Should you throw exception if secMgrCfg is nullptr? If secMgrCfg is nullptr, m_userPass will not be set and the authorize() will throw exception. So, why not throw exception here if secMgrCfg is nullptr?

File: system/security/plugins/singleuserSecurity/singleUserSecurity.hpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2013 HPCC Systems®.
```
Comment (wangkx) [inline system/security/plugins/singleuserSecurity/singleUserSecurity.hpp]: 2018

File: esp/platform/espcfg.cpp [hunk 3]
```diff
@@ -281,10 +282,41 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
         startPerformanceMonitor(m_cfg->getPropInt("@perfReportDelay", 60)*1000);
 #endif
 
-        //get the local computer name:              
+        if (m_inputs->hasProp("SingleUsePass"))
+        {
+            StringBuffer espuserpass;
+            m_inputs->getProp("SingleUsePass", espuserpass);
+            StringBuffer xpath;
+            xpath.setf("SecurityManagers/SecurityManager[@type=\"SingleUserSecurityManager\"]SingleUserSecurityManager/");
```
Comment (wangkx) [inline esp/platform/espcfg.cpp]: ?

**Review by wangkx — COMMENTED** [review]
@rpastrana most of my comments are minors.

File: system/security/plugins/singleuserSecurity/singleUserSecurity.cpp (hunk 1 above)
Comment (rpastrana) [reply system/security/plugins/singleuserSecurity/singleUserSecurity.cpp]: I agree, the problem is we tend not to throw in a constructor

**Comment by rpastrana** [comment]
@wangkx please review latest commit.

File: system/security/plugins/singleuserSecurity/CMakeLists.txt [hunk 4]
```diff
@@ -38,8 +38,8 @@ include_directories (
          ${HPCC_SOURCE_DIR}/system/jlib
          ${HPCC_SOURCE_DIR}/esp/platform
          ${HPCC_SOURCE_DIR}/dali/base
-         ${APR_INCLUDE_DIR}
-         ${APRUTIL_INCLUDE_DIR}
+         #${APR_INCLUDE_DIR}
+         #${APRUTIL_INCLUDE_DIR}
```
Comment (wangkx) [inline system/security/plugins/singleuserSecurity/CMakeLists.txt]: Why not remove those 2 lines?

File: system/security/plugins/singleuserSecurity/CMakeLists.txt [hunk 5]
```diff
@@ -49,8 +49,8 @@ install ( TARGETS singleUserSecurity RUNTIME DESTINATION ${EXEC_DIR} LIBRARY DES
 target_link_libraries ( singleUserSecurity
          jlib
          dalibase
-         ${APR_LIBRARIES}
-         ${APRUTIL_LIBRARIES}
+         #${APR_LIBRARIES}
+         #${APRUTIL_LIBRARIES}
```
Comment (wangkx) [inline system/security/plugins/singleuserSecurity/CMakeLists.txt]: Why not remove those 2 lines?

File: system/security/plugins/singleuserSecurity/singleUserSecurity.cpp [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+        return authmap;
+    }
+
+    IAuthMap * createFeatureMap(IPropertyTree * authconfig) override
+    {
+        CAuthMap* feature_authmap = new CAuthMap(this);
+
+        Owned<IPropertyTreeIterator> feature_iter;
+        feature_iter.setown(authconfig->getElements(".//Feature"));
```
[the remark this answers is not in the document]
Comment (wangkx) [reply system/security/plugins/singleuserSecurity/singleUserSecurity.cpp]: Forgot this comment?

File: system/security/plugins/singleuserSecurity/singleUserSecurity.cpp [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        ForEach(*feature_iter)
+        {
+            IPropertyTree *feature = NULL;
+            feature = &feature_iter->query();
```
[the remark this answers is not in the document]
Comment (wangkx) [reply system/security/plugins/singleuserSecurity/singleUserSecurity.cpp]: Forgot this one?

**Review by wangkx — COMMENTED** [review]
@rpastrana back to you. For some reason, my 2 comments on system/security/plugins/singleuserSecurity/singleUserSecurity.cpp (Line 102 and Line 106) are not shown here. Please check.

**Comment by rpastrana** [comment]
@wangkx one more review please

**Review by wangkx — APPROVED** [review]
@rpastrana looks good.

**Comment by rpastrana** [comment]
@richardkchapman please merge

**Comment by rpastrana** [comment]
@wangkx please review. Added pass encryption. thanks.

**Review by wangkx — COMMENTED** [review]
@rpastrana minor: 2 2016 s

**Comment by rpastrana** [comment]
@wangkx last code review please

**Comment by wangkx** [comment]
@rpastrana looks good.

**Pull request merged**